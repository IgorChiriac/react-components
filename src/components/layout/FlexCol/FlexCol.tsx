import React, { FC } from 'react';
import styled, { css } from 'styled-components';
import grid, { TGridBreakpoints } from '../../../constants/grid';

export type TAlignSelf = 'flex-start' | 'center' | 'flex-end' | 'stretch' | 'auto';
export type TColWidth = number | 'fill' | 'auto' | 'hidden';

export interface IFlexColProps {
  align?: TAlignSelf;
  cols?: number;
  gutter?: number;
  l?: TColWidth;
  m?: TColWidth;
  s?: TColWidth;
  xl?: TColWidth;
  xs?: TColWidth;
}

export interface IFlexCol extends React.HTMLAttributes<HTMLDivElement>, IFlexColProps {}

const genHidden = (breakpoint: TGridBreakpoints) => css`
  @media (min-width: ${grid.breakpoints[breakpoint]}px) {
    display: none;
  }
`;

const genFillWidth = (breakpoint: TGridBreakpoints) => css`
  @media (min-width: ${grid.breakpoints[breakpoint]}px) {
    display: block;
    flex-basis: 0;
    flex-grow: 1;
    max-width: 100%;
  }
`;

const genAutoWidth = (breakpoint: TGridBreakpoints) => css`
  @media (min-width: ${grid.breakpoints[breakpoint]}px) {
    display: block;
    flex: 0 0 auto;
    width: auto;
    max-width: none;
  }
`;

const genRelativeWidth = (breakpoint: TGridBreakpoints, colWidth: TColWidth, cols: IFlexColProps['cols']) => css`
  @media (min-width: ${grid.breakpoints[breakpoint]}px) {
    display: block;
    ${typeof colWidth === 'number' && typeof cols === 'number'
      ? `
    flex: 0 0 ${(colWidth / cols) * 100}%;
    max-width: ${(colWidth / cols) * 100}%;
    `
      : null}
  }
`;

const getWidth = (breakpoint: TGridBreakpoints, colWidth: TColWidth, cols: IFlexColProps['cols']) => {
  if (colWidth === 'hidden') {
    return genHidden(breakpoint);
  }
  if (colWidth === 'auto') {
    return genAutoWidth(breakpoint);
  }
  if (colWidth === 'fill') {
    return genFillWidth(breakpoint);
  }
  return genRelativeWidth(breakpoint, colWidth, cols);
};

const StyledFlexCol = styled.div<IFlexCol>`
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: ${({ gutter }) => gutter}px;
  padding-left: ${({ gutter }) => gutter}px;
  align-self: ${({ align }) => align};
  ${props =>
    Object.keys(grid.breakpoints)
      .sort((a, b) => {
        // Note: TS infers `a` and `b` as strings, hence the type assertion here.
        return grid.breakpoints[a as TGridBreakpoints] - grid.breakpoints[b as TGridBreakpoints];
      })
      .filter(breakpoint => Object.keys(props).includes(breakpoint))
      .map(breakpoint => {
        // Note: TS infers `breakpoint` as string again... hence more type assertion here.
        const colWidth = props[breakpoint as TGridBreakpoints] || 'auto';
        return getWidth(breakpoint as TGridBreakpoints, colWidth, props.cols);
      })}
`;

const FlexCol: FC<IFlexColProps> = props => <StyledFlexCol {...props} />;

FlexCol.defaultProps = {
  align: 'auto',
};

export default FlexCol;
