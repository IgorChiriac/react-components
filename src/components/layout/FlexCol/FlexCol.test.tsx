import React from 'react';
import { render } from '@testing-library/react';
import FlexCol from './FlexCol';

describe('<FlexCol />', () => {
  it('renders a FlexCol component', () => {
    const { container } = render(<FlexCol />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
