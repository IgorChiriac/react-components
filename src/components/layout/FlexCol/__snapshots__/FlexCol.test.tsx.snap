// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlexCol /> renders a FlexCol component 1`] = `
.c0 {
  position: relative;
  width: 100%;
  min-height: 1px;
  padding-right: px;
  padding-left: px;
  -webkit-align-self: flex-start;
  -ms-flex-item-align: start;
  align-self: flex-start;
}

<div
  class="c0"
/>
`;
